### DISCUSSION TITLE
How to deal with 'Error: Dynamic server usage: force-dynamic'?

### DISCUSSION THREAD

**Discussion by sebpowell** [vercel/next.js#52911] [Help]
### Summary

Hello, I'm working on an eCommerce site to which I want to add the ability for users to switch to a different currency. When the user does this, I want to store their preference as a cookie. 

This is how I'm currently approaching it:

```TSX
export default async function RootLayout({
  children,
}: {
  children: ReactNode;
}) {
  const store = await fetchStoreSettings(GetStore());

  const currencyIso = await cookies().get(Cookies.STORE_SETTINGS_COOKIE);

  const selectedCurrency = store?.currencies.find(
    (currency) => currency.iso === currencyIso?.value
  ) || { iso: DEFAULT_CURRENCY_ISO, conversionRate: 1 };

  return (
    <LayoutMain
      preloadedState={{
        settingsSlice: {
          selectedCurrency,
        },
      }}
    >
      {children}
    </LayoutMain>
  );
```

Here, I'm getting the currency ISO code stored in the user cookie, and then using that to work out the conversion rate set for it in a set of values being returned by my API. The selected value is then passed to a component that is wrapped in a Redux provider so I can easily share the values in all components.

This works without any problems. However, when clicking on a link that takes me to a dynamic route (in this case, `/products/[productSlug]`), I get the following error:

<img width="987" alt="Screenshot 2023-07-19 at 22 39 08" src="https://github.com/vercel/next.js/assets/1786366/22fba110-f7cc-4322-af38-9e7e6b2bfda1">

This only happens when navigating from one page to another – if I refresh the page, it loads without any problems. This is a simplified version of the code that generates that page:

```TSX
export const generateStaticParams = async () => {
  const productVariants = await fetchProductVariants();

  const slugs =
    productVariants.items
      .filter((variant) => variant.fields.product.fields !== undefined)
      .map((productVariant) => ({
        slug: productVariant.fields.slug,
      })) || [];

  return slugs;
};

export default async function PageProduct({
  params,
}: {
  params: { slug: ProductVariant["slug"] };
}) {
  return (
      <ProductPage />
  );
}
```
I believe the issue is related to what is being discussed here: https://github.com/vercel/next.js/issues/49373

And I was able to resolve the issue by changing the above to something like this (which is the recommended fix):

```TSX
export const dynamic = "force-dynamic";

const staticParams = async () => {
  const productVariants = await fetchProductVariants();

  const slugs =
    productVariants.items
      .filter((variant) => variant.fields.product.fields !== undefined)
      .map((productVariant) => ({
        slug: productVariant.fields.slug,
      })) || [];

  return slugs;
};

export const generateStaticParams =
  process.env.NODE_ENV === "production" ? staticParams : undefined;

export default async function PageProduct({
  params,
}: {
  params: { slug: ProductVariant["slug"] };
}) {
  return (
      <ProductPage />
  );
}
```

The issue with is however is that I *think* this is now forcing the page to be rendered dynamically, meaning it doesn't load quite as fast as it used to before I implemented this feature (as everything was statically generated). 

So I suppose what I'm wondering is; is there a way for me to solve the above so that the page is still statically generated? I'm guessing not, as the page, if statically generated has no knowledge of the user's preferences. Which I guess means the only other thing I could do is load the cookies in on the client side – which I have also tried, and that works fine, the only downside is that it takes sometimes a couple of seconds before the cookies are read, meaning the user will see prices in the default currency before seeing them in the one they previously selected (this is only really a 'problem' if they refresh the page). 

I hope that makes sense, and curious to hear people's thoughts. Thank you in advance.

**Comment by Jordaneisenburger** [comment]
I have the following which works great without `generateStaticParams` but obviously I'd like to enable it for better performance. It partially works with `generateStaticParams` set but then logged in customers won't see their specific message because cookies are empty. So my next thought would be to get the cookie client side inside the `Price` component and pass it to the promise coming from the parent so that we still fetch data on the server as recommended. But then I would need to mark the Price component as 'use client' which prevents it from being an async component meaning I can't await the promise anymore... everytime when I think I'm close I hit a road block. 

I also tried placing the fetch inside a useEffect but this will cause the content to flash :(

```jsx
// productpage

import { Price } from '@/components/price/price';
import { Suspense } from 'react';
import { cookies } from 'next/headers';

export function generateStaticParams() {
    return [];
}

const getData = async () => {
    const cookie = cookies().get('commerce-auth-token');

    const { data, errors } = await fetch(
        'https://blabla.com/graphql',
        {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${cookie?.value}`
            },
            body: JSON.stringify({
                query: `
                query getCustomer {
                    customer {
                        firstname
                        email
                        lastname
                    }
                }
              `
            })
        }
    )
        .then((res) => res.json())
        .then((result) => result);

    if (errors) {
        return null;
    }

    return {
        ...data.customer
    };
};

export default function Product({ params }: { params: any }) {
    const userData = getData();

    return (
        <main className="container  mx-auto grid max-w-7xl grid-cols-2 gap-2">
            <div className="flex h-[500px] items-center justify-center bg-gray-600 text-white">
                Image gallery
            </div>
            <div className="bg-gray-200 p-8">
                <h1 className="text-3xl font-bold tracking-tight text-gray-900">
                    Product title
                </h1>
                <div className="mt-3">
                    <Suspense fallback="loading...">
                        <Price promise={userData} />
                    </Suspense>
                </div>

                <div className="space-y-6 text-base text-gray-700">
                    The Zip Tote Basket is the perfect midpoint between shopping
                    tote and comfy backpack. With convertible straps, you can
                    hand carry, should sling, or backpack this convenient and
                    spacious bag. The zip top and durable canvas construction
                    keeps your goods protected for all-day use.
                </div>
            </div>
            <div className="flex h-[300px] items-center justify-center bg-gray-600 text-white">
                Product specs
            </div>
        </main>
    );
}
```

```jsx
// price component

export const Price = async ({ promise }: { promise: any }) => {
    const customer = await promise;

    return customer ? (
        <div>
            <p className="text-3xl tracking-tight text-gray-900">
                $140 special for {customer.firstname}
            </p>
        </div>
    ) : (
        <div>
            <p className="text-3xl tracking-tight text-gray-900">$200</p>
        </div>
    );
};


```

**Reply by sebpowell** [reply]
@Jordaneisenburger ah I see – when I was trying the client side approach, I was trying something a little different, I don't know if it would help. 

Essentially, the approach involves getting the cookies inside the component like so:

```TSX
import Cookies from "js-cookie"

const Price = () => {

  const currencyCookie = Cookies.get("CURRENCY_PREFERENCE_COOKIE")

  const [isLoading, setLoading] = useState(true);  

  useEffect(() => {
    setLoading(false)
  }, [currencyCookie])
  
  return (
    isLoading <>Loading...</> : `${currencyCookie || "£"}123`
  )
}
```

Looking at your example above, I'm guessing you could adjust this to be something like so:

```TSX
import Cookies from "js-cookie"

const Price = () => {
  const cookie = Cookies.get("commerce-auth-token")

  const [isLoading, setLoading] = useState(true);  

  const [data, setData] = useState();  

  useEffect(() => {
       const { data, errors } = await fetch(
        'https://blabla.com/graphql',
        {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${cookie?.value}`
            },
            body: JSON.stringify({
                query: `
                query getCustomer {
                    customer {
                        firstname
                        email
                        lastname
                    }
                }
              `
            })
        }
    )
        .then((res) => res.json())
        .then((result) => {
          setLoading(false)
          setData(result)
     });
  }, [cookie])
  
  return (
    isLoading <>Loading...</> : `data comes here...`
  )
}
```
I don't know if that potentially solves it for you?

**Reply by sebpowell** [reply]
@Jordaneisenburger sorry just saw your last comment – yes, the above would cause content flash, but I think that's inevitable if client side cookies are used.

**Reply by Jordaneisenburger** [reply]
exactly so the logged-in user would always (quickly) see not logged in prices before seeing his actuall price :/

**Reply by sebpowell** [reply]
@Jordaneisenburger – ok yep, I see – same thing I noticed then. I guess you could hide it behind a loading indicator as per the above example, but I get that this is not ideal (I wasn't overly happy about it either), and also I guess this is not great for SEO either.

**Reply by Jordaneisenburger** [reply]
For now I guess best way is to keep it Dynamic but put `prefetch={true}` on all `<Link />` elements so that all subsequent page views for a logged in customer are appearing to be fast.
